src/conn/conn_tiered.c

```diff
@@ -120,23 +120,158 @@ __tier_storage_remove_local(WT_SESSION_IMPL *session, const char *uri, bool forc
  */
 static int
-__tier_storage_copy(WT_SESSION_IMPL *session)
+__tier_flush_meta(
+  WT_SESSION_IMPL *session, WT_TIERED *tiered, const char *local_uri, const char *obj_uri)
 {
+    WT_DATA_HANDLE *dhandle;
+    WT_DECL_ITEM(buf);
+    WT_DECL_RET;
+    uint64_t now;
+    char *newconfig, *obj_value;
+    const char *cfg[3] = {NULL, NULL, NULL};
+    bool tracking;
+
+    tracking = false;
+    WT_RET(__wt_scr_alloc(session, 512, &buf));
+    dhandle = &tiered->iface;
+
+    newconfig = NULL;
+    WT_ERR(__wt_meta_track_on(session));
+    tracking = true;
+
+    WT_ERR(__wt_session_get_dhandle(session, dhandle->name, NULL, NULL, WT_DHANDLE_EXCLUSIVE));
     /*
-     * Walk the work queue and copy file:<name> to shared storage object:<name>. Walk a tiered
-     * table's tiers array and copy it to any tier that allows WT_TIERS_OP_FLUSH.
+     * Once the flush call succeeds we want to first remove the file: entry from the metadata and
+     * then update the object: metadata to indicate the flush is complete.
      */
-    /* XXX: We don't want to call this here, it is just to quiet the compiler that this function
-     * can return NULL. So it is a placeholder until we have real content here.
+    WT_ERR(__wt_metadata_remove(session, local_uri));
+    WT_ERR(__wt_metadata_search(session, obj_uri, &obj_value));
+    __wt_seconds(session, &now);
+    WT_ERR(__wt_buf_fmt(session, buf, "flush=%" PRIu64, now));
+    cfg[0] = obj_value;
+    cfg[1] = buf->mem;
+    WT_ERR(__wt_config_collapse(session, cfg, &newconfig));
+    WT_ERR(__wt_metadata_update(session, obj_uri, newconfig));
+    WT_ERR(__wt_meta_track_off(session, true, ret != 0));
+    tracking = false;
+
+err:
+    __wt_free(session, newconfig);
+    WT_TRET(__wt_session_release_dhandle(session));
+    __wt_scr_free(session, &buf);
+    if (tracking)
+        WT_TRET(__wt_meta_track_off(session, true, ret != 0));
+    return (ret);
+}
+
+/*
+ * __wt_tier_do_flush --
+ *     Perform one iteration of copying newly flushed objects to the shared storage.
+ */
+int
+__wt_tier_do_flush(
+  WT_SESSION_IMPL *session, WT_TIERED *tiered, const char *local_uri, const char *obj_uri)
+{
+    WT_DECL_RET;
+    WT_FILE_SYSTEM *bucket_fs;
+    WT_STORAGE_SOURCE *storage_source;
+    const char *local_name, *obj_name;
+
+    storage_source = tiered->bstorage->storage_source;
+    bucket_fs = tiered->bstorage->file_system;
+
+    local_name = local_uri;
+    WT_PREFIX_SKIP_REQUIRED(session, local_name, "file:");
+    obj_name = obj_uri;
+    WT_PREFIX_SKIP_REQUIRED(session, obj_name, "object:");
+
+    /* This call make take a while, and may fail due to network timeout. */
+    WT_RET(storage_source->ss_flush(
+      storage_source, &session->iface, bucket_fs, local_name, obj_name, NULL));
+
+#if 1
+    WT_WITH_CHECKPOINT_LOCK(session,
+      WT_WITH_SCHEMA_LOCK(session, ret = __tier_flush_meta(session, tiered, local_uri, obj_uri)));
+    WT_RET(ret);
+#else
```

Comment:
Do we need the code in the `#else` block here? Is this (and similar) `#if 1` blocks stuff for debugging WT-7567?